collection/zset/zset.go

```diff
@@ -0,0 +1,879 @@
+//
+// The returned elements are ordered by score, from highest to lowest.
+// Elements with the same score are ordered in reverse lexicographical ordering.
+//
+// This function won't panic even when the given rank out of range.
+//
+// RevRange is the replacement of ZREVRANGE command of redis.
+func (z *Float64Set) RevRange(start, stop int) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	// Convert negative rank
+	if start < 0 {
+		start = z.list.length + start
+	}
+	if stop < 0 {
+		stop = z.list.length + stop
+	}
+
+	var res []Float64Node
+	x := z.list.GetNodeByRank(z.list.length - start) // 0-based rank -> 1-based rank
+	for x != nil && start <= stop {
+		start++
+		res = append(res, Float64Node{
+			Score: x.score,
+			Value: x.value,
+		})
+		x = x.prev
+	}
+	return res
+}
+
+// RevRangeByScore returns all the elements in the sorted set with a
+// score between max and min (including elements with score equal to max or min).
+// The elements are considered to be ordered from high to low scores.
+//
+// RevRangeByScore is the replacement of ZREVRANGEBYSCORE command of redis.
+func (z *Float64Set) RevRangeByScore(max, min float64) []Float64Node {
+	return z.RevRangeByScoreWithOpt(max, min, Float64RangeOpt{})
+}
+
+func (z *Float64Set) RevRangeByScoreWithOpt(max, min float64, opt Float64RangeOpt) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	var res []Float64Node
+	x := z.list.LastInRange(min, max, opt)
+	for x != nil && (x.score > min || (!opt.ExcludeMin && x.score == min)) {
+		res = append(res, Float64Node{
+			Score: x.score,
+			Value: x.value,
+		})
+		x = x.prev
+	}
+	return res
+}
+
+// RemoveRangeByRank removes all elements in the sorted set stored with rank
+// between start and stop.
+// Both start and stop are 0-based, they can also be negative numbers indicating
+// offsets from the end of the sorted set, with -1 being the last element of the sorted set,
+// and so on.
+//
+// RemoveRangeByRank is the replacement of ZREMRANGEBYRANK command of redis.
+func (z *Float64Set) RemoveRangeByRank(start, stop int) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	// Convert negative rank
+	if start < 0 {
+		start = z.list.length + start
+	}
+	if stop < 0 {
+		stop = z.list.length + stop
+	}
+
+	return z.list.DeleteRangeByRank(start+1, stop+1, z.dict) // 0-based rank -> 1-based rank
+}
+
+// RemoveRangeByScore removes all elements in the sorted set stored with a score
+// between min and max (including elements with score equal to min or max).
+//
+// RemoveRangeByScore is the replacement of ZREMRANGEBYSCORE command of redis.
+func (z *Float64Set) RemoveRangeByScore(min, max float64) []Float64Node {
+	return z.RevRangeByScoreWithOpt(min, max, Float64RangeOpt{})
+}
+
+func (z *Float64Set) RemoveRangeByScoreWithOpt(min, max float64, opt Float64RangeOpt) []Float64Node {
+	z.mu.RLock()
+	defer z.mu.RUnlock()
+
+	return z.list.DeleteRangeByScore(min, max, opt, z.dict)
+}
+
+//
+// skip list implementation
```

Comment:
I am not sure maybe we could move the skiplist implementation into another file? This file is too big, more than 800 lines.

Reply:
Yes, maybe move all reusable parts into other files is a better idea.